Implement Daemonset history

janetkuo opened the pull request:
~Depends on #45867 (the 1st commit, ignore it when reviewing)~ (already merged)

1. Update DaemonSet API for rollback and history
    * Add revisionHistoryLimit (default 10), collisionCount, and validation code
    * Add daemonset-controller-hash label, and deprecate templateGeneration
2. Implement DaemonSet history logic in controller 
   * Create controllerrevisions (history) and label pods with template hash for both RollingUpdate and OnDelete update strategy
   * Clean up old, non-live history based on revisionHistoryLimit
   * Remove duplicate controllerrevisions (the ones with the same template) and relabel their pods
   * Update RBAC to allow DaemonSet controller to manage controllerrevisions
   * In DaemonSet controller unit tests, create new pods with hash labels

Ref https://github.com/kubernetes/community/pull/527/ and https://github.com/kubernetes/community/pull/594

@kubernetes/sig-apps-api-reviews @kubernetes/sig-apps-pr-reviews @erictune @kow3ns @lukaszo @kargakis 

---

TODOs:
- [x] API changes
  - [x] (maybe) Remove rollback subresource if we decide to do client-side rollback 
- [x] deployment controller 
  - [x] controller revision
    - [x] owner ref (claim & adoption)
    - [x] history reconstruct (put revision number, hash collision avoidance)
    - [x] de-dup history and relabel pods
    - [x] compare ds template with history 
  - [x] hash labels (put it in controller revision, pods, and maybe deployment)
  - [x] clean up old history 
  - [x] Rename status.uniquifier when we reach consensus in #44774 
- [x] e2e tests 
- [x] unit tests 
  - [x] daemoncontroller_test.go 
  - [x] update_test.go 
  - [x] ~(maybe) storage_test.go // if we do server side rollback~

kubectl part is in #46144

--- 

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
```

janetkuo committed:
Update DaemonSet API for rollback and history

1. Add revisionHistoryLimit (default 10), collisionCount, and validation code
2. Add daemonset-controller-hash label, and deprecate templateGeneration

janetkuo committed:
Implement DaemonSet history logic in controller

1. Create controllerrevisions (history) and label pods with template
   hash for both RollingUpdate and OnDelete update strategy
2. Clean up old, non-live history based on revisionHistoryLimit
3. Remove duplicate controllerrevisions (the ones with the same template)
   and relabel their pods
4. Update RBAC to allow DaemonSet controller to manage
   controllerrevisions
5. In DaemonSet controller unit tests, create new pods with hash labels

janetkuo committed:
Test both strategies in all daemonSet controller unit tests

kargakis commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -516,10 +546,16 @@ type DaemonSet struct {
+	// DefaultDaemonSetUniqueLabelKey is the default label key that is added
+	// to existing DaemonSet pods to distinguish between old and new
+	// DaemonSet pods during DaemonSet template updates.
+	DefaultDaemonSetUniqueLabelKey string = "daemonset-controller-revision"
```
`daemonset.kubernetes.io/revision` to match Deployments and StatefulSets

janetkuo replied:
Isn't Deployment using `pod-template-hash`?

kargakis replied:
Oh, you are not going to use pod-template-hash too? Are you going to relabel every time on rollbacks?

janetkuo replied:
Not using `pod-template-hash` intentionally so that we differentiate histories from different controllers. 

No need to relabel, each history will have a fixed hash label and the label is used by DS to compare the DS's template with the history and to label DS's pods

kargakis commented on pkg/client/listers/extensions/v1beta1/daemonset_expansion.go:
```diff
@@ -76,3 +78,38 @@ func (s *daemonSetLister) GetPodDaemonSets(pod *v1.Pod) ([]*v1beta1.DaemonSet, e
+		return nil, fmt.Errorf("no DaemonSet found for ControllerRevision %s because it has no labels", history.Name)
+	}
+
+	// TODO: MODIFY THIS METHOD so that it checks for the podTemplateSpecHash label
```
Not sure we want to do this anymore.

janetkuo replied:
Removed

kargakis commented on pkg/client/listers/extensions/v1beta1/daemonset_expansion.go:
```diff
@@ -76,3 +78,38 @@ func (s *daemonSetLister) GetPodDaemonSets(pod *v1.Pod) ([]*v1beta1.DaemonSet, e
 	return daemonSets, nil
 }
+
+// GetHistoryDaemonSets returns a list of DaemonSets that potentially
```
It doesn't seem that you care about the list, return directly the CR that matches the controller ref.

janetkuo replied:
This DS list is for orphan controller history adoption.

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -273,6 +289,138 @@ func (dsc *DaemonSetsController) getPodDaemonSets(pod *v1.Pod) []*extensions.Dae
+			return
+		}
+		glog.V(4).Infof("ControllerRevision %s added.", history.Name)
+		dsc.enqueueDaemonSet(ds)
```
Why do we want to enqueue in this case?

janetkuo replied:
Removed

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -273,6 +289,138 @@ func (dsc *DaemonSetsController) getPodDaemonSets(pod *v1.Pod) []*extensions.Dae
+}
+
+// updateHistory figures out what DaemonSet(s) manage a ControllerRevision when the ControllerRevision
+// is updated and wake them up. If the anything of the ControllerRevision have changed, we need to
```
Isn't ControllerRevision read-only?

janetkuo replied:
From the proposal, only its `.data` cannot be updated. We need to at least be able to update its owner ref, otherwise we can't handle orphan / adoption / garbage collection.

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -453,11 +601,11 @@ func (dsc *DaemonSetsController) updateNode(old, cur interface{}) {
 }
 
-// getNodesToDaemonSetPods returns a map from nodes to daemon pods (corresponding to ds) running on the nodes.
+// getDaemonPods returnsdaemon pods owned by the given ds.
```
missing space

janetkuo replied:
Fixed

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -64,20 +85,275 @@ func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet) error {
+	var currentHistories []*apps.ControllerRevision
+	histories, err = dsc.controlledHistories(ds)
+	if err != nil {
+		return
```
return nil, nil, err

kargakis replied:
same for the rest of the return statements

janetkuo replied:
Done

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -64,20 +85,275 @@ func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet) error {
+	for _, history := range histories {
+		// Add the unique label if it's not already added to the history
+		// We use history name instead of computing hash, so that we don't need to worry about hash collision
+		if _, ok := history.Labels[extensions.DefaultDaemonSetUniqueLabelKey]; !ok {
```
Why do we need to update these with an extra label?

janetkuo replied:
For adoption of manually created controller history. Can remove this if we don't want to support it.

kargakis replied:
The DS should already be able to adopt any existing history based on the selector set by the user. Do you want to avoid being adopted by a StatefulSet once you orphan?

janetkuo replied:
The `history` here are only those which are selected and controlled (owner ref) by the DS. Adding the label to the history here because current (new) history needs the unique label so that DS can label its pods.

kargakis replied:
These are metadata labels, do you mean to add them in .data instead?

janetkuo replied:
No, we can't update `.data`.

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -64,20 +85,275 @@ func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet) error {
+
+	history, err = dsc.kubeClient.AppsV1beta1().ControllerRevisions(ds.Namespace).Create(history)
+	if errors.IsAlreadyExists(err) {
+		// TODO: Is it okay to get from historyLister?
```
Yes, it is. If you can't find it, you return the error and retry later?

janetkuo replied:
But you won't find it in cache when you hit this case, right?

kargakis replied:
Why not?

kargakis replied:
And if you don't find it, you already retry.

janetkuo replied:
> Why not?

Because we only create new history (call `snapshot`) when an existing one isn't found.

kargakis replied:
You just got a AlreadyExists error - I haven't looked closely, what am i missing?

janetkuo replied:
AlreadyExists is from API server. If we saw this in cache (when trying to find current history of a DaemonSet) we wouldn't try to create it and instead just reuse the existing one.

kargakis replied:
Note that even the API server is using its own cache so this may still be a stale read.

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -64,20 +85,275 @@ func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet) error {
+		}
+
+		// Handle name collisions between different history
+		// TODO: Is it okay to get from dsLister?
```
You need to validate decrementing, then doing this should be fine.

janetkuo replied:
I feel it's safer to get from API server, and we only do this call upon name collisions so the performance impact will be low

janetkuo committed:
Autogen: run hack/update-all.sh

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -322,7 +322,6 @@ func (dsc *DaemonSetsController) addHistory(obj interface{}) {
 			return
```
The comment of this handler is not 100% true, this is also called on controller manager restarts.

janetkuo replied:
Comment updated

janetkuo commented:
Addressed comments, and add some major changes (see the commits starts with "FIXUP"):
1. Removed /rollback sub-resource and related API changes, we'll do it client-side
2. Label pods with hash labels and creates history for both RollingUpdate and OnDelete (originally only RollingUpdate). This is because we need to avoid unnecessary restarts of DaemonSet pods when a DaemonSet is switched from OnDelete to RollingUpdate (can't compare with history and without labeling the pod with hash)
3. Fixed unit test failure and updated unit tests to make sure we test both update strategies in all cases

janetkuo commented:
Addressed comments, and add some major changes (see the commits starts with "FIXUP"):
1. Removed /rollback sub-resource and related API changes, we'll do it client-side
1. Label pods with hash labels and creates history for both RollingUpdate and OnDelete (originally only RollingUpdate). This is because we need to avoid unnecessary restarts of DaemonSet pods when a DaemonSet is switched from OnDelete to RollingUpdate (can't compare with history and without labeling the pod with hash)
1. Fixed unit test failure and updated unit tests to make sure we test both update strategies in all cases 
1. Clean up old history 
1. Allow controller history revision to be updated (need this for the rollback case)

janetkuo committed:
Verify histories and pods in DaemonSet e2e test

janetkuo commented:
pull-kubernetes-federation-e2e-gce is flaky: #45978

janetkuo commented:
I'll need to rename `status.uniquifier` when we reach consensus in #44774 (most likely to `collisionCount`)

janetkuo commented:
@k8s-bot pull-kubernetes-unit test this
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

janetkuo commented:
Rebased and renamed `uniquifier` to `collisionCount`

erictune commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -579,23 +735,37 @@ func (dsc *DaemonSetsController) manage(ds *extensions.DaemonSet) error {
 		}
 	}
-	errors := dsc.syncNodes(ds, podsToDelete, nodesNeedingDaemonPods)
+
```
The previous blocks in this function have good comments.  This new code would benefit from them too.

janetkuo replied:
Add comment for below code block

janetkuo commented on pkg/apis/extensions/validation/validation.go:
```diff
@@ -134,13 +134,18 @@ func validateDaemonSetStatus(status *extensions.DaemonSetStatus, fldPath *field.
 	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(status.UpdatedNumberScheduled), fldPath.Child("updatedNumberScheduled"))...)
 	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(status.NumberAvailable), fldPath.Child("numberAvailable"))...)
 	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(status.NumberUnavailable), fldPath.Child("numberUnavailable"))...)
+	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(status.NumberUnavailable), fldPath.Child("numberUnavailable"))...)
```
Yes, must be caused by rebase 😞

janetkuo commented on pkg/apis/extensions/types.go:
```diff
@@ -443,10 +443,17 @@ type DaemonSetSpec struct {
+
+	// The number of old history to retain to allow rollback.
+	// This is a pointer to distinguish between explicit zero and not specified.
+	// Defaults to 2.
```
Picked 2 because Deployment chose 2. How about 5?

kargakis replied:
It's unlike that you will need more than three but I guess it depends on your update cadence. I would expect stuff that run as Deployments will release sooner?

smarterclayton replied:
I'd be ok with 5

janetkuo replied:
Changed to 10 based on offline discussion with @erictune

erictune commented:
The code for adoption of controller histories (CHs) by Daemon Sets is tricky.  Is this is just the same logic as adoption of pods by controllers?   I think I understand some of the use cases for pod adoption, but I'm not sure I understand any of the use cases for CH adoption.  On one hand I can see a case for saying  "controllers always try to adopt orphaned but label-matching objects, regardless of type".  OTOH, I don't want to generalize from a single instance.   And Pods are different from a data-only objects.

erictune commented:
In the controllerHistory proposal, 
https://github.com/kubernetes/community/pull/594, it says *controller will store a unique, monotonically increasing revision number*, but a hash (not monotonically increasing) is stored in the label `daemonset-controller-revision`.   The names don't match (revision vs hash).  But also, I wonder if small integer, not-hash might be a more user friendly choice?  (It does get messy if you try to adopt a CH to a controller with a different integer sequence, but maybe adoption is less important?

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -273,6 +289,138 @@ func (dsc *DaemonSetsController) getPodDaemonSets(pod *v1.Pod) []*extensions.Dae
+			return
+		}
+		glog.V(4).Infof("ControllerRevision %s added.", history.Name)
+		return
```
I think you want to add the requeue back, especially if the rollout depends in the existence of history (not sure yet if it does).

janetkuo replied:
Although the rollout depends on the existence of history, we don't need to requeue on addHistory. 

If the step for creating a new history failed during a rollout, the controller will error out and requeue the DaemonSet (rate limited); if that step succeeded, it'll continue the rollout (creating new pods). If any history is added, the existing DaemonSet won't be affected as long its template isn't changed (trigger a rollout).

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -579,23 +735,37 @@ func (dsc *DaemonSetsController) manage(ds *extensions.DaemonSet) error {
+		return "", fmt.Errorf("failed to construct revisions of DaemonSet: %v", err)
+	}
+
+	err = dsc.cleanupHistory(ds, old, nodeToDaemonPods)
```
I only moved history cleanup for Deployments so we could unblock Deployments broken by hash collisions. I think it's strictly more correct to cleanup at steady state, i.e., not before or during upgrades and I am planning on moving it back to its original place now that we have the hashing collision avoidance mechanism in place.

janetkuo replied:
@kargakis are you suggesting moving `cleanupHistory` to the last step of `syncDaemonSet` (after rolling update)? That makes sense, will do so.

janetkuo replied:
Done

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -64,20 +76,318 @@ func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet) error {
+}
+
+// computeHash returns a hash value calculated from pod template and a collisionCount to avoid hash collision
+// TODO: this is copied from deployment controller util, will move to a shared library for all controllers
```
Can be collapsed now

janetkuo replied:
Done

janetkuo commented:
> In the controllerHistory proposal, 
https://github.com/kubernetes/community/pull/594, it says *controller will store a unique, monotonically increasing revision number*, but a hash (not monotonically increasing) is stored in the label `daemonset-controller-revision`.   The names don't match (revision vs hash).  But also, I wonder if small integer, not-hash might be a more user friendly choice?  (It does get messy if you try to adopt a CH to a controller with a different integer sequence, but maybe adoption is less important?   

@erictune the "unique, monotonically increasing revision number" is stored in each history's `.revision` field. We want the unique label value to be unique, not be confused with revision, and consistent with other controllers (i.e. deployments and statefulsets) -- therefore hash is used. We can rename the hash label key to something like "daemon-controller-hash".

smarterclayton commented:
API changes LGTM modulo eric's comment about saving more.

janetkuo commented on pkg/apis/extensions/validation/validation.go:
```diff
@@ -134,13 +134,18 @@ func validateDaemonSetStatus(status *extensions.DaemonSetStatus, fldPath *field.
 func ValidateDaemonSetStatusUpdate(ds, oldDS *extensions.DaemonSet) field.ErrorList {
 	allErrs := apivalidation.ValidateObjectMetaUpdate(&ds.ObjectMeta, &oldDS.ObjectMeta, field.NewPath("metadata"))
 	allErrs = append(allErrs, validateDaemonSetStatus(&ds.Status, field.NewPath("status"))...)
+	// TODO: validate that status.uniquifier is never decremented
```
This makes sure that controller doesn't set it to a smaller number because of stale cache.

janetkuo commented:
Rebased and addressed comments:

1. API: changed unique label key to "daemonset-controller-hash"
1. API: changed default revisionHistoryLimit to 10 (based on offline discussion with @erictune) 
1. Bug fix: copy history labels from ds template labels (instead of from ds labels)
1. Move computeHash to a shared library for controllers 
1. Clean up DaemonSet history after (instead of before) rolling update

janetkuo commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

smarterclayton commented:
Other comments?

erictune commented on pkg/apis/extensions/validation/validation.go:
```diff
@@ -134,13 +134,23 @@ func validateDaemonSetStatus(status *extensions.DaemonSetStatus, fldPath *field.
 	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(status.NumberAvailable), fldPath.Child("numberAvailable"))...)
 	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(status.NumberUnavailable), fldPath.Child("numberUnavailable"))...)
+	if status.CollisionCount != nil {
+		allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(*status.CollisionCount), fldPath.Child("collisionCount"))...)
```
Not sure how necessary this and the not-decrementing change are below.  The controller can be trusted to use the collision counter correctly, I think?

kargakis replied:
Not when it uses a cache.

erictune commented on pkg/controller/controller_utils.go:
```diff
@@ -980,3 +981,18 @@ func WaitForCacheSync(controllerName string, stopCh <-chan struct{}, cacheSyncs
+		podTemplateSpecHasher.Write(collisionCountBytes)
+	}
+
+	return podTemplateSpecHasher.Sum32()
```
The 64bit collision count gets hashed down to 32 bits. 🤷‍♂️

erictune reviewed (commented):
LGTM.  

A few comments that don't need to be addressed.

erictune commented:
Does the following work with this code, and if so, would an e2e test for it be useful (in a follow up PR):

1. create DS
2. update template - rollout proceeds only a little ways
3. roll back by updating template to be the same as the original template.  existing pods not yet updated  in step 2 should be relabeled without restarting

janetkuo commented:
Rebased and squashed commits. Will address the above comments in follow-up PRs. Applying tag.

janetkuo commented:
@smarterclayton would you approve this?

smarterclayton commented:
/approve

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -64,20 +73,311 @@ func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet) error {
+			toUpdate.Labels = make(map[string]string)
+		}
+		toUpdate.Labels[extensions.DefaultDaemonSetUniqueLabelKey] = cur.Labels[extensions.DefaultDaemonSetUniqueLabelKey]
+		_, err = dsc.kubeClient.ExtensionsV1beta1().DaemonSets(ds.Namespace).UpdateStatus(toUpdate)
```
We are going to invalidate label updates during status updates at some point in the future and this will break.

kargakis replied:
https://github.com/kubernetes/kubernetes/issues/20978

janetkuo replied:
Will follow up

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -64,20 +73,311 @@ func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet) error {
+	if err != nil {
+		return fmt.Errorf("couldn't get node to daemon pod mapping for daemon set %q: %v", ds.Name, err)
+	}
+	_, old, err := dsc.constructHistory(ds)
```
I am wondering if it's going to be cleaner to construct the history once and then pass the references in `manage` and `cleanupHistory`, instead of reconstructing it for each one of them.

janetkuo replied:
Sounds good, will follow up

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -64,20 +73,311 @@ func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet) error {
+			}
+		}
+	}
+	// Label ds with current history's unique label as well
```
Why do we need this actually?

janetkuo replied:
It's easier for us to check if pods are up-to-date without looking though all history again when calculating DS status

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -825,21 +991,28 @@ func (dsc *DaemonSetsController) syncDaemonSet(key string) error {
 		}
 	}
 
+	err = dsc.cleanupHistory(ds)
```
Has `rollingUpdate` finished the whole update at this point?

janetkuo replied:
Yes, `rollingUpdate` calls `syncNodes` which creates & deletes pods and waits for those operations to complete

kargakis replied:
Does this mean that a stuck update will hold a worker forever?

janetkuo replied:
Create & delete pods won't wait forever. If any creation/deletion fails, it collects the error, and returns all errors.

kargakis replied:
Successful creates / deletes doesn't mean complete rollout so `rollingUpdate` hasn't actually finished the whole rollout.

janetkuo replied:
I see. Is the concern about not killing old history when it has a pod that is terminating?

kargakis replied:
I don't see this as an issue for now but if we standarize on having cleanup being executed at the end of a rollout, we should fix this. As I said elsewhere, I am going to rollback the Deployment case now that we have collisionCount (remember that cleanup was executed at the end before and I think it's strictly correct to do so).

kargakis replied:
As an example, Jenkins is cleaning up old builds only after the new builds are complete.

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -152,6 +160,14 @@ func NewDaemonSetsController(daemonSetInformer extensionsinformers.DaemonSetInfo
+		DeleteFunc: dsc.deleteHistory,
+	})
+	dsc.historyLister = historyInformer.Lister()
+	dsc.historyStoreSynced = historyInformer.Informer().HasSynced
```
Don't you want to wait for the history cache to be synced?

janetkuo replied:
Yes, will add in `WaitForCacheSync`

janetkuo commented:
Rebased. Bumping priority to (hopefully) avoid another rebase...

janetkuo commented:
This blocks #46144

janetkuo commented:
@k8s-bot pull-kubernetes-unit test this
